src/vsr/message_header.zig

```diff
@@ -0,0 +1,1430 @@
+const std = @import("std");
+const assert = std.debug.assert;
+
+const constants = @import("../constants.zig");
+const stdx = @import("../stdx.zig");
+const vsr = @import("../vsr.zig");
+const Command = vsr.Command;
+const Operation = vsr.Operation;
+const schema = @import("../lsm/schema.zig");
+
+/// Aegis128 (used by checksum) uses hardware accelerated AES via inline asm which isn't
+/// available at comptime. Use a hard-coded value instead and verify via a test.
+const checksum_body_empty: u128 = 0x49F174618255402DE6E7E3C40D60CC83;
+
+test "empty checksum" {
+    assert(checksum_body_empty == vsr.checksum(&.{}));
+}
+
+/// Network message and journal entry header:
+/// We reuse the same header for both so that prepare messages from the primary can simply be
+/// journalled as is by the backups without requiring any further modification.
+pub const Header = extern struct {
+    /// A checksum covering only the remainder of this header.
+    /// This allows the header to be trusted without having to recv() or read() the associated body.
+    /// This checksum is enough to uniquely identify a network message or journal entry.
+    checksum: u128,
+
+    /// A checksum covering only the associated body after this header.
+    checksum_body: u128,
+
+    /// Reserved for future use by AEAD.
+    reserved_nonce: u128,
+
+    /// The cluster number binds intention into the header, so that a client or replica can indicate
+    /// the cluster it believes it is speaking to, instead of accidentally talking to the wrong
+    /// cluster (for example, staging vs production).
+    cluster: u32,
+
+    /// The size of the Header structure (always), plus any associated body.
+    size: u32,
+
+    /// The cluster reconfiguration epoch number (for future use).
+    epoch: u32,
+
+    /// Every message sent from one replica to another contains the sending replica's current view.
+    /// A `u32` allows for a minimum lifetime of 136 years at a rate of one view change per second.
+    view: u32,
+
+    /// The version of the protocol implementation that originated this message.
+    version: u16,
+
+    /// The Viewstamped Replication protocol command for this message.
+    command: Command,
+
+    /// The index of the replica in the cluster configuration array that authored this message.
+    /// This identifies only the ultimate author because messages may be forwarded amongst replicas.
+    replica: u8,
+
+    /// Reserved for future use by the header frame (i.e. to be shared by all message types).
+    reserved_frame: [12]u8,
+
+    /// This data's schema is different depending on the `Header.command`.
+    /// (No default value – `Header`s should not be constructed directly.)
+    reserved_command: [176]u8,
+
+    comptime {
+        assert(@sizeOf(Header) == 256);
+        assert(stdx.no_padding(Header));
+        assert(@offsetOf(Header, "reserved_command") % @alignOf(u128) == 0);
```

Comment:
```suggestion
        assert(@offsetOf(Header, "reserved_command") % 16 == 0);
```

`@alignOf(u128) == 8`

Reply:
I wonder if we should patch `no_padding` to check also that all fields are aligned at least to their sizeof.